### PR TITLE
ensure no forward slashes in names for HDF5-based backends

### LINKED ISSUE #7943: '/' in variable name causes to_netcdf() to crash
### What happened?

If variable name includes a '/' then to_netcdf crashes

### What did you expect to happen?

_No response_

### Minimal Complete Verifiable Example

```Python
import pandas as pd
df = pd.read_csv('slashCSV.csv')
index = ['temperature']
df = df.set_index(index)
xarr = df.to_xarray()
xarr.to_netcdf("new.nc")


- slashCSV.csv contents:
PASS/FAIL,temperature
PASS, 10
FAIL, 20
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

```Python
File "/opt/homebrew/lib/python3.11/site-packages/xarray/backends/netCDF4_.py", line 56, in __init__
    array = self.get_array()
            ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/xarray/backends/netCDF4_.py", line 83, in get_array
    variable = ds.variables[self.variable_name]
               ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
KeyError: 'PASS/FAIL'
```

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.2 (main, Mar 24 2023, 00:16:47) [Clang 14.0.0 (clang-1400.0.29.202)]
python-bits: 64
OS: Darwin
OS-release: 22.5.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.1

xarray: 2023.2.0
pandas: 1.5.3
numpy: 1.24.2
scipy: 1.10.1
netCDF4: 1.6.3
pydap: None
h5netcdf: 1.1.0
h5py: 3.8.0
Nio: None
zarr: 2.14.2
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.3.6
cfgrib: 0.9.10.3
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: None
cartopy: None
seaborn: None
numbagg: None
fsspec: 2023.3.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 65.6.3
pip: 23.1.1
conda: None
pytest: None
mypy: None
IPython: None
sphinx: None
</details>

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] Closes #7943
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/backends/netCDF4_.py [hunk 1]
```diff
@@ -194,6 +194,13 @@ def _nc4_require_group(ds, group, mode, create_group=_netcdf4_create_group):
         return ds
 
 
+def _ensure_no_forward_slash_in_name(name):
+    if "/" in name:
```
Comment (headtr1ck) [inline xarray/backends/netCDF4_.py:198]: I'm not sure what the type of name is at this stage. Could it be `None` or non-str `Hashable`?

**Review by headtr1ck — APPROVED** [review]
Looks good. Thanks!

File: xarray/backends/netCDF4_.py (hunk 1 above)
Comment (kmuehlbauer) [reply xarray/backends/netCDF4_.py:198]: Ah, good catch. If that's the case, we'll have to adapt that check. I'll try to get behind it.

Comment (kmuehlbauer) [reply xarray/backends/netCDF4_.py:198]: @headtr1ck From what I can tell, the names should already be strings at that point of the backend code. So this should be safe.

File: xarray/backends/netCDF4_.py [hunk 2]
```diff
@@ -194,6 +194,13 @@ def _nc4_require_group(ds, group, mode, create_group=_netcdf4_create_group):
...
+        raise ValueError(f"Forward slashes '/' are not allowed in name {name!r} for "  
+                         "HDF5-based files ('netcdf4'/'h5netcdf'). Forward slashes "
+                         "are used as hierarchy-separators.")
```
Comment (jhamman) [inline xarray/backends/netCDF4_.py:201]:
```suggestion
        raise ValueError(f"Forward slashes '/' are not allowed in variable names (got {name!r}). "  
                         "Forward slashes are used as are used as hierarchy-separators for "
                         "HDF5-based files ('netcdf4'/'h5netcdf').")
```

**Review by jhamman — APPROVED** [review]
Minor suggestion but this looks great.

File: xarray/backends/netCDF4_.py (hunk 2 above)
Comment (kmuehlbauer) [reply xarray/backends/netCDF4_.py]: Thanks @jhamman, I've added dimensions, as they are treated likewise.

**Pull request merged**